src/components/App.jsx

```diff
@@ -0,0 +1,209 @@
+import ProtectedRoute from "./ProtectedRoute";
+import SavedNews from "./SavedNews";
+import SigninModal from "./SigninModal";
+import SignupModal from "./SignupModal";
+
+export default function App() {
+  const [searchState, setSearchState] = useState({
+    results: [],
+    keyword: "",
+    articlesAvail: 0,
+    articlesShown: 3,
+    loading: false,
+    nothingFound: false,
+  });
+  const [userState, setUserState] = useState({
+    loggedIn: false,
+    email: "",
+    username: "",
+    savedNews: [],
+  });
+  const [activeModal, setActiveModal] = useState("");
+  const [protectedDestination, setProtectedDestination] = useState("");
+  const navigate = useNavigate();
+
+  const api = new Api();
+
+  const newsApi = new NewsApi({
+    baseUrl: "https://nomoreparties.co/news/v2/everything",
+    apiKey: "a16de474931b4e5a83f83ad53ba3df69",
+  });
+
+  useEffect(() => {
+    if (protectedDestination !== "") setActiveModal("signin");
+  }, [protectedDestination]);
+
+  useEffect(() => {
+    document.addEventListener("keydown", (e) => {
+      if (e.key === "Escape") closeActiveModal();
+    });
+
+    document.querySelectorAll(".modal").forEach((modal) => {
+      modal.addEventListener("click", (e) => {
+        if (e.target.classList[0] === "modal") closeActiveModal();
+      });
+    });
+  }, []);
+
+  const handleSearchSubmit = (query) => {
+    setSearchState((currState) => ({
+      ...currState,
+      loading: true,
+      nothingFound: false,
+      articlesShown: 3,
+    }));
+    newsApi
+      .getNewsArticles(query)
+      .then((res) => {
+        setSearchState((currState) => ({
+          ...currState,
+          keyword: query,
+          articlesAvail: res.totalResults,
+          results: res.articles.filter(
+            (article) => article.title !== "[Removed]"
+          ),
+          nothingFound: res.totalResults <= 0 ? true : false,
+        }));
+      })
+      .catch((err) => console.error(err))
+      .finally(() =>
+        setSearchState((currState) => ({ ...currState, loading: false }))
+      );
+  };
+
+  const addSavedArticle = (newArticle) => {
+    api.addArticle(newArticle).then((res) => {
+      const updatedSavedNews = [...userState.savedNews, res];
+      setUserState({
+        ...userState,
+        savedNews: updatedSavedNews,
+      });
+    });
+  };
+
+  const removeSavedArticle = (id) => {
+    api.removeArticle(id).then((res) => {
+      setUserState({
+        ...userState,
+        savedNews: [
+          ...userState.savedNews.filter((article) => article.id != res.id),
+        ],
+      });
+    });
+  };
+
+  const handleSignin = (values) => {
+    authorize(values)
+      .then((res) => {
+        setUserState({
+          ...userState,
+          loggedIn: true,
+          username: res.username,
+          email: res.email,
+        });
+      })
+      .catch((err) => console.error(err));
+    navigate(protectedDestination || "/");
+    closeActiveModal();
```

Comment:
You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server), and if you change the button text, you could see the change as well. **Please, correct it everywhere in the code**